### PR TITLE
Add RepositoryPlugin interface for registering snapshot repositories

### PR DESCRIPTION
**Pull request by rjernst**
Repository plugins currently use a lot of custom classes like
RepositoryName and RepositorySettings in order to use guice to construct
repository implementations. But repositories now only really need their
settings to be constructed. Anything else they need (eg a cloud client)
can be constructed within the plugin, instead of via guice.

This change makes repository plugins use the new pull model. It removes
guice from the construction of Repository objects (no more child
injectors) and also from all repository plugins.

### CODE REVIEW

File: plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java [hunk 1]
```diff
@@ -1,353 +0,0 @@
...
-import java.io.IOException;
-
-import static org.elasticsearch.repositories.s3.S3Repository.Repositories;
-import static org.elasticsearch.repositories.s3.S3Repository.Repository;
-import static org.elasticsearch.repositories.s3.S3Repository.getValue;
-import static org.hamcrest.Matchers.containsString;
-import static org.hamcrest.Matchers.is;
-import static org.hamcrest.Matchers.isEmptyString;
-
-public class RepositoryS3SettingsTests extends ESTestCase {
```
Comment (dadoonet) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java]: Why did you remove those tests?

If possible I'd like to keep them as I introduced them after we found a bunch of errors in our code when it came to dealing with settings coming from `cloud.aws`, or `cloud.aws.s3` or `repositories.s3` or within repository instance settings.

Comment (rjernst) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java]: > Why did you remove those tests?

It was partially accidental. It had looked like the test was just testing Repository settings which I removed. I can add it back, but I also question what it is really testing. It seems to try to be exhaustive, but in a very error prone way (easy to miss new settings). Is this really anything more than a test of the logic in the helper method to find settings from the repo and fallback to node settings?

File: core/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 2]
```diff
@@ -346,55 +343,44 @@ public Repository repository(String repository) {
...
         try {
-            ModulesBuilder modules = new ModulesBuilder();
-            RepositoryName name = new RepositoryName(repositoryMetaData.type(), repositoryMetaData.name());
-            modules.add(b -> {
-                b.bind(RepositoryName.class).toInstance(name);
-                typesRegistry.bindType(b, repositoryMetaData.type());
-                b.bind(RepositorySettings.class).toInstance(new RepositorySettings(settings, repositoryMetaData.settings()));
-            });
-
-            repositoryInjector = modules.createChildInjector(injector);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: since we own that code :( can we remove that createChildInjector method as well or throw UOE?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/AzureRepositoryModule.java [hunk 3]
```diff
@@ -1,57 +0,0 @@
...
- * Azure Module
- *
- * <ul>
- * <li>If needed this module will bind azure repository service by default
- * to AzureStorageServiceImpl.</li>
- * </ul>
- *
- * @see org.elasticsearch.cloud.azure.storage.AzureStorageServiceImpl
- */
-public class AzureRepositoryModule extends AbstractModule {
```
Comment (s1monw) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/AzureRepositoryModule.java]: oh god thank you!

File: plugins/repository-azure/src/main/java/org/elasticsearch/plugin/repository/azure/AzureRepositoryPlugin.java [hunk 4]
```diff
@@ -20,42 +20,36 @@
...
     @Override
-    public Collection<Module> nodeModules() {
-        return Collections.singletonList((Module) new AzureRepositoryModule(settings));
-    }
-
-    public void onModule(RepositoriesModule module) {
-        logger.debug("registering repository type [{}]", AzureRepository.TYPE);
-        module.registerRepository(AzureRepository.TYPE, AzureRepository.class);
+    public Map<String, Repository.Factory> getRepositories(Environment env) {
+        return Collections.singletonMap(AzureRepository.TYPE,
```
Comment (s1monw) [inline plugins/repository-azure/src/main/java/org/elasticsearch/plugin/repository/azure/AzureRepositoryPlugin.java]: are you sure you wanna call new I mean it could create an object? 👊

**Comment by s1monw** [comment]
oh man this is so good! LGTM

File: plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java (hunk 1 above)
Comment (dadoonet) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java]: I agree that most of tests are testing this logic. It just helped me to check when I'm adding a new setting that everything works as I'd expect. 

So it's the way I'm developing in repository plugins: 

1. add a setting
2. unit test that it works as I'm expecting
3. implement the logic with it.

But I agree that might be too many tests here. I don't have a strong opinion on that TBH.

Only `testInvalidChunkBufferSizeRepositorySettings` is actually a more valuable test that we should keep around.

Comment (rjernst) [inline plugins/repository-s3/src/test/java/org/elasticsearch/cloud/aws/RepositoryS3SettingsTests.java]: I added back the chunk/buffer size tests, and also added back a simple test of `getValue`

File: core/src/main/java/org/elasticsearch/repositories/RepositoriesService.java (hunk 2 above)
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/repositories/RepositoriesService.java]: I will open a follow up PR for this (guice is tricky to decode...what can be removed and what cannot...).
